internal/stores/hostdb.go

```diff
@@ -181,3 +195,236 @@ func NewJSONHostDB(dir string) (*JSONHostDB, modules.ConsensusChangeID, error) {
+		// higher priority.
+		Host      []byte    `gorm:"index:idx_host:2"`
+		Timestamp time.Time `gorm:"index:idx_host:1"`
+	}
+
+	// consensusInfo defines table which stores the latest consensus info
+	// known to the hostdb. It should only ever contain a single entry with
+	// the consensusInfoID primary key.
+	consensusInfo struct {
+		ID   uint8 `gorm:"primaryKey"`
+		CCID []byte
+	}
+)
+
+// Host converts a host into a hostdb.Host.
+func (h host) Host() hostdb.Host {
+	hdbHost := hostdb.Host{}
+	copy(hdbHost.PublicKey[:], h.PublicKey)
+	for _, a := range h.Announcements {
+		hdbHost.Announcements = append(hdbHost.Announcements, a.Announcement())
+	}
+	for _, i := range h.Interactions {
+		hdbHost.Interactions = append(hdbHost.Interactions, i.Interaction())
+	}
+	return hdbHost
+}
+
+// Announcement converts a host into a hostdb.Announcement.
+func (a announcement) Announcement() hostdb.Announcement {
+	hostdbAnnouncement := hostdb.Announcement{
+		Index: consensus.ChainIndex{
+			Height: a.BlockHeight,
+		},
+		Timestamp:  a.Timestamp,
+		NetAddress: a.NetAddress,
+	}
+	copy(hostdbAnnouncement.Index.ID[:], a.BlockID)
+	return hostdbAnnouncement
+}
+
+// Interaction converts an interaction into a hostdb.Interaction.
+func (i interaction) Interaction() hostdb.Interaction {
+	return hostdb.Interaction{
+		Timestamp: i.Timestamp,
+		Type:      i.Type,
+		Result:    i.Result,
+	}
+}
+
+// NewSQLHostDB uses a given Dialector to connect to a SQL database.  NOTE: Only
+// pass migrate=true for the first instance of SQLHostDB if you connect via the
+// same Dialector multiple times.
+func NewSQLHostDB(conn gorm.Dialector, migrate bool) (*SQLHostDB, modules.ConsensusChangeID, error) {
+	db, err := gorm.Open(conn, &gorm.Config{})
+	if err != nil {
+		return nil, modules.ConsensusChangeID{}, err
+	}
+
+	if migrate {
+		// Create the tables.
+		if err := db.AutoMigrate(&host{}); err != nil {
+			return nil, modules.ConsensusChangeID{}, err
+		}
+		if err := db.AutoMigrate(&interaction{}); err != nil {
+			return nil, modules.ConsensusChangeID{}, err
+		}
+		if err := db.AutoMigrate(&announcement{}); err != nil {
+			return nil, modules.ConsensusChangeID{}, err
+		}
+		if err := db.AutoMigrate(&consensusInfo{}); err != nil {
+			return nil, modules.ConsensusChangeID{}, err
+		}
+	}
+
+	// Get latest consensus change ID or init db.
+	var ci consensusInfo
+	err = db.Where(&consensusInfo{ID: consensusInfoID}).
+		Attrs(consensusInfo{
+			ID:   consensusInfoID,
+			CCID: modules.ConsensusChangeBeginning[:],
+		}).
+		FirstOrCreate(&ci).Error
+	if err != nil {
+		return nil, modules.ConsensusChangeID{}, err
+	}
+	var ccid modules.ConsensusChangeID
+	copy(ccid[:], ci.CCID)
+
+	return &SQLHostDB{
+		staticDB: db,
+	}, ccid, nil
+}
+
+// Host returns information about a host.
+func (db *SQLHostDB) Host(hostKey consensus.PublicKey) (hostdb.Host, error) {
+	var h host
+	tx := db.staticDB.Where(&host{PublicKey: hostKey[:]}).
+		Preload("Interactions").
```

Comment:
what does `Preload` do?

Reply:
By default, the interactions and announcements are empty slices since SQL doesn't have slices.
Those are actually values from the interactions and announcements tables which contain foreign keys.

Preload will query those tables and populate the slices in the Go type. So it's sort of a convenience method for us not to have to manually select and create the slices.

Reply:
I believe it eagerly loads the relationship so you get hydrated/populated fields from `Find`